### PR TITLE
Add releaseNotes.json Creation Step in Upload Packs

### PR DESCRIPTION
**Pull request by amshamah419**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: internal

## Description
Adding step which generates and updates `releaseNotes.json` file in the GCP bucket

### CODE REVIEW

File: Tests/Marketplace/marketplace_services.py [hunk 1]
```diff
@@ -565,6 +568,92 @@ def upload_to_storage(self, zip_pack_path, latest_version, storage_bucket, overr
...
+                latest_int = self.major_minor_rev(latest)
+                _version = str(latest).replace('.md', '')
+                version = str(_version).replace('_', '.')
+                if not expected_release_version_int == latest_int:
+                    raise Exception(f"Conflict between version found in metadata ({expected_release_version_str}) "
+                                    f"and latest version found in ReleaseNotes dir ({version})")
+                    # sys.exit(1)
+
+                # We need to retrieve previous releaseNotes.json here
+                latest_gcp_rn_path = os.path.join(GCPConfig.STORAGE_BASE_PATH, self._pack_name, _version, '.json')
```
Comment (IkaDemisto) [inline Tests/Marketplace/marketplace_services.py:629]: Not clear. Why do you need it?

File: Tests/Marketplace/marketplace_services.py [hunk 2]
```diff
@@ -565,6 +568,92 @@ def upload_to_storage(self, zip_pack_path, latest_version, storage_bucket, overr
...
+        return int(major), int(minor), int(rev)
+
+    def prepare_release_notes(self, storage_bucket):
+        """Need to implement the changelog.md parsing and changelog.json creation after design is
+        finalized.
+
+        """
+        task_status = False
+        try:
+            metadata_path = os.path.join(self._pack_path, Pack.METADATA)
```
Comment (IkaDemisto) [inline Tests/Marketplace/marketplace_services.py:594]: Can remove it... This already done in `load_user_metadata`. Use `self.current_version` property for it. If you need additional data, create new property.

File: Tests/Marketplace/marketplace_services.py [hunk 3]
```diff
@@ -565,6 +568,92 @@ def upload_to_storage(self, zip_pack_path, latest_version, storage_bucket, overr
...
+        :param path: Bool to indicate if searching within a path or string
+        :return: Tuple of integers.
+        """
+        if path:
+            major, minor, rev = re.search('(\d+)_(\d+)_(\d+)', version).groups()
+        else:
+            major, minor, rev = re.search('(\d+)\.(\d+)\.(\d+)', version).groups()
+        return int(major), int(minor), int(rev)
+
+    def prepare_release_notes(self, storage_bucket):
```
Comment (IkaDemisto) [inline Tests/Marketplace/marketplace_services.py:587]: Where is creation of `changelog.json`?

File: Tests/Marketplace/marketplace_services.py [hunk 4]
```diff
@@ -565,6 +568,92 @@ def upload_to_storage(self, zip_pack_path, latest_version, storage_bucket, overr
...
+                _version = str(latest).replace('.md', '')
+                version = str(_version).replace('_', '.')
+                if not expected_release_version_int == latest_int:
+                    raise Exception(f"Conflict between version found in metadata ({expected_release_version_str}) "
+                                    f"and latest version found in ReleaseNotes dir ({version})")
+                    # sys.exit(1)
+
+                # We need to retrieve previous releaseNotes.json here
+                latest_gcp_rn_path = os.path.join(GCPConfig.STORAGE_BASE_PATH, self._pack_name, _version, '.json')
+                existing_rn_file = storage_bucket.blob(prefix=latest_gcp_rn_path)
```
Comment (IkaDemisto) [inline Tests/Marketplace/marketplace_services.py:630]: Wrong logic. You have all versions of release notes in `index` (`changelog.json` file).
Similar to what @liorblob did with metadata versions.

File: Tests/Marketplace/upload_packs.py [hunk 5]
```diff
@@ -504,8 +504,11 @@ def main():
             pack.cleanup()
             continue
 
-        # todo finish implementation of release notes
-        # pack.parse_release_notes()
+        task_status = pack.prepare_release_notes(storage_bucket)
+        if not task_status:
+            pack.status = PackStatus.FAILED_RELEASE_NOTES.name
```
Comment (IkaDemisto) [inline Tests/Marketplace/upload_packs.py:509]: If you planning to create `changelog.json` in that function, so the error message should be different.

File: Tests/Marketplace/marketplace_services.py [hunk 6]
```diff
@@ -565,6 +568,77 @@ def upload_to_storage(self, zip_pack_path, latest_version, storage_bucket, overr
...
+                                             'released': datetime.utcnow().strftime(Metadata.DATE_FORMAT)}
+                        changelog[latest_release_notes] = version_changelog
+                        with open(os.path.join(self._pack_path, 'changelog.json'), "w") as f:
+                            json.dump(changelog, f)
+
+                else:
+                    print_warning(f"No ReleaseNotes directory was found for pack: {self._pack_name}")
+
+            else:
+                print_warning(f"No release notes found for: {self._pack_name}")
```
Comment (IkaDemisto) [inline Tests/Marketplace/marketplace_services.py:632]: This is edge case that should be handled properly. Most of packs are have initial version `1.0.0` and don't have `changelog.json`.  I think we need to discuss it.

File: Tests/Marketplace/marketplace_services.py [hunk 7]
```diff
@@ -565,6 +568,76 @@ def upload_to_storage(self, zip_pack_path, latest_version, storage_bucket, overr
...
+        """Need to implement the changelog.md parsing and changelog.json creation after design is
+        finalized.
+
+        """
+        task_status = False
+        try:
+            release_notes_dir = os.path.join(self._pack_path, Pack.RELEASE_NOTES)
+            if os.path.exists(os.path.join(index_folder_path, self._pack_name, 'changelog.json')):
+                print_warning(f"Found Changelog for: {self._pack_name}")
+                changelog_path = os.path.join(self._pack_path, Pack.CHANGELOG_JSON)
```
[the remark this answers is not in the document]
Comment (amshamah419) [reply Tests/Marketplace/marketplace_services.py:597]: I just read it from the index, then dump a new file into the `_pack_path`

[the remark this answers is not in the document]
Comment (IkaDemisto) [reply Tests/Marketplace/marketplace_services.py:597]: Not sure where it is happening 😄

[the remark this answers is not in the document]
Comment (amshamah419) [reply Tests/Marketplace/marketplace_services.py:597]: That's because it shouldn't have worked, but we have a changelog in the repo so it wasn't failing like it should have lol. Fixed now ;)

[the remark this answers is not in the document]
Comment (IkaDemisto) [reply Tests/Marketplace/marketplace_services.py]: Yea, I think only in HelloWorld

**Comment by IkaDemisto** [comment]
@amshamah419 @yuvalbenshalom up to you if you want to remove `-o` flag in this PR.

File: Tests/Marketplace/marketplace_services.py [hunk 8]
```diff
@@ -565,6 +579,73 @@ def upload_to_storage(self, zip_pack_path, latest_version, storage_bucket, overr
             print_error(f"Failed in uploading {self._pack_name} pack to gcs. Additional info:\n {e}")
             return task_status, True
 
+    def prepare_release_notes(self, index_folder_path):
+        """Need to implement the changelog.md parsing and changelog.json creation after design is
+        finalized.
```
Comment (yuvalbenshalom) [inline Tests/Marketplace/marketplace_services.py:584]: is that a final comment?

File: Tests/Marketplace/Tests/marketplace_services_test.py [hunk 9]
```diff
@@ -225,3 +225,27 @@ def test_not_existing_changelog_json(self, mocker, dummy_pack):
...
+    def dummy_pack(self):
+        """ dummy pack fixture
+        """
+        dummy_path = os.path.join(os.path.dirname(os.path.abspath(__file__)), "test_data")
+        sample_pack = Pack(pack_name="TestPack", pack_path=dummy_path)
+        sample_pack.description = 'Sample description'
+        sample_pack.current_version = '1.0.0'
+        return sample_pack
+
+    def test_prepare_release_notes(self, mocker, dummy_pack):
```
Comment (yuvalbenshalom) [inline Tests/Marketplace/Tests/marketplace_services_test.py:245]: we need more tests to Pack.prepare_release_notes. that will cover:
1) when version already exists in changelo
2) Version mismatch detected in release notes and metadata.
3) when change log exist with version 1.0.0  and a new version 1.0.1 is added.(all good)

and add standard test comments with, start, when, expected result

File: dev-requirements-py3.txt [hunk 10]
```diff
@@ -1,7 +1,7 @@
 flake8==3.7.9
 bandit==1.6.2
 demisto-py==2.0.9
-demisto-sdk==1.0.0
+demisto-sdk==1.0.1
```
Comment (yuvalbenshalom) [inline dev-requirements-py3.txt:4]: should this bee here

Comment (amshamah419) [reply dev-requirements-py3.txt:4]: I'm not sure. I know we were updating the build to use the new SDK.

I'll check.

File: Tests/Marketplace/Tests/marketplace_services_test.py [hunk 11]
```diff
@@ -287,8 +287,8 @@ def test_prepare_release_notes_upgrade_version(self, mocker, dummy_pack):
             }
         }'''
         new_changelog_file = '# This is a new release for 2.0.2'
-        mocker.patch('builtins.open', mock_open(read_data=original_changelog))
-        mocker.patch('builtins.open', mock_open(read_data=new_changelog_file))
+        mocker.patch('builtins.open', mock_open(read_data=original_changelog), new_callable=mock_open())
+        mocker.patch('builtins.open', mock_open(read_data=new_changelog_file), new_callable=mock_open())
```
Comment (liorblob) [inline Tests/Marketplace/Tests/marketplace_services_test.py:291]: You can chain mocks by giving a list to the `side_effect` argument

File: dev-requirements-py3.txt (hunk 10 above)
Comment (amshamah419) [reply dev-requirements-py3.txt:4]: It's 1.0.1 now but will be bumped to 1.0.2 today.
